packages/backend/src/tokens/request.ts

```diff
@@ -56,85 +45,287 @@ function assertSignInUrlFormatAndOrigin(_signInUrl: string, origin: string) {
+  const { cookies, headers, searchParams, derivedRequestUrl } = buildRequest(options.request);
 
-  const ruleOptions = {
+  const authenticateContext = {
     ...options,
     ...loadOptionsFromHeaders(headers),
     ...loadOptionsFromCookies(cookies),
     searchParams,
-  } satisfies InterstitialRuleOptions;
+    derivedRequestUrl,
+  };
+
+  const devBrowserToken = searchParams?.get('__clerk_db_jwt') || cookies('__clerk_db_jwt') || '';
+  const handshakeToken = searchParams?.get('__clerk_handshake') || cookies('__clerk_handshake') || '';
+
+  assertValidSecretKey(authenticateContext.secretKey);
+
+  if (authenticateContext.isSatellite) {
+    assertSignInUrlExists(authenticateContext.signInUrl, authenticateContext.secretKey);
+    if (authenticateContext.signInUrl && authenticateContext.origin) {
+      assertSignInUrlFormatAndOrigin(authenticateContext.signInUrl, authenticateContext.origin);
+    }
+    assertProxyUrlOrDomain(authenticateContext.proxyUrl || authenticateContext.domain);
+  }
+
+  function buildRedirectToHandshake() {
+    const redirectUrl = new URL(derivedRequestUrl);
+    redirectUrl.searchParams.delete('__clerk_db_jwt');
+    const frontendApiNoProtocol = pk.frontendApi.replace(/http(s)?:\/\//, '');
 
-  assertValidSecretKey(ruleOptions.secretKey);
+    const url = new URL(`https://${frontendApiNoProtocol}/v1/client/handshake`);
+    url.searchParams.append('redirect_url', redirectUrl?.href || '');
 
-  if (ruleOptions.isSatellite) {
-    assertSignInUrlExists(ruleOptions.signInUrl, ruleOptions.secretKey);
-    if (ruleOptions.signInUrl && ruleOptions.origin) {
-      assertSignInUrlFormatAndOrigin(ruleOptions.signInUrl, ruleOptions.origin);
+    if (pk?.instanceType === 'development' && devBrowserToken) {
+      url.searchParams.append('__clerk_db_jwt', devBrowserToken);
     }
-    assertProxyUrlOrDomain(ruleOptions.proxyUrl || ruleOptions.domain);
+
+    return new Headers({ location: url.href });
   }
 
+  async function resolveHandshake() {
+    const headers = new Headers({
+      'Access-Control-Allow-Origin': 'null',
+      'Access-Control-Allow-Credentials': 'true',
+    });
+
+    const cookiesToSet = JSON.parse(atob(handshakeToken)) as string[];
+
+    let sessionToken = '';
+    cookiesToSet.forEach((x: string) => {
+      headers.append('Set-Cookie', x);
+      if (x.startsWith('__session=')) {
+        sessionToken = x.split(';')[0].substring(10);
+      }
+    });
+
+    if (instanceType === 'development') {
+      const newUrl = new URL(authenticateContext.derivedRequestUrl);
+      newUrl.searchParams.delete('__clerk_handshake');
+      newUrl.searchParams.delete('__clerk_help');
+      headers.append('Location', newUrl.toString());
+    }
+
+    if (sessionToken === '') {
+      return signedOut(authenticateContext, AuthErrorReason.SessionTokenMissing, '', headers);
+    }
+
+    let verifyResult: JwtPayload;
+
+    try {
+      verifyResult = await verifyToken(sessionToken, authenticateContext);
```

Comment:
TODO: for a follow up PR: `verifyToken` should return result, error to avoid controlling the flow with via an exception. The code will read much better.

cc: @nikosdouvlis this affects the V5 API exports